server/src/main/java/io/deephaven/server/object/ObjectServiceGrpcImpl.java

```diff
@@ -26,28 +21,206 @@
+                            "Data message sent before Connect message");
+                }
+                Data data = request.getData();
+                List<SessionState.ExportObject<Object>> referenceObjects = data.getExportedReferencesList().stream()
+                        .map(typedTicket -> ticketRouter.resolve(session, typedTicket.getTicket(), "ticket"))
+                        .collect(Collectors.toList());
+                List<SessionState.ExportObject<Object>> requireObjects = new ArrayList<>(referenceObjects);
+                requireObjects.add(object);
+                runOrEnqueue(requireObjects, () -> {
+                    Object[] objs = referenceObjects.stream().map(ExportObject::get).toArray();
+                    messageStream.onData(data.getPayload().asReadOnlyByteBuffer(), objs);
+                });
+            }
+        }
+
+        private final Queue<StreamOperation> operations = new ConcurrentLinkedQueue<>();
+        private final AtomicBoolean running = new AtomicBoolean(false);
+
+        // These methods are intended to roughly behave like SerializingExecutor(directExecutor()) in that only one can
+        // be running at a time, and will be started on the current thread, with the distinction that submitted work
+        // will continue off-thread and will signal when it is finished
+        private void runOrEnqueue(StreamOperation runnable) {
+            runOrEnqueue(Collections.emptyList(), runnable);
+        }
+
+        private void runOrEnqueue(Collection<? extends ExportObject<?>> additExports, StreamOperation operation) {
+            List<ExportObject<?>> requirements = new ArrayList<>(additExports);
+            requirements.add(object);
+
+            StreamOperation runnable = () -> {
+                session.nonExport()
+                        .onErrorHandler(this::onError)
+                        .require(requirements)
+                        .submit(() -> {
+                            // Run the specified work. Note that we're not concerned about exceptions, the stream will
+                            // be dead (via onError) and won't be used again.
+                            try {
+                                operation.run();
+                            } catch (ObjectType.ObjectCommunicationException e) {
+                                throw Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT,
+                                        "Error performing MessageStream operation");
+                            }
+
+                            // Set running to false (it must be true at this time) so that any new work can race being
+                            // added
+                            running.set(false);
+                            doWork();
+                        });
+            };
+
+            // gRPC guarantees we can't race enqueuing
+            operations.add(runnable);
```

Comment:
The items that go into `operations` are always this `session.nonExport()...` `Runnable` - and that does _not_ need to be `StreamOperation` since it doesn't throw that error; as such, I think we should change:

```java
        private final Queue<StreamOperation> operations = new ConcurrentLinkedQueue<>();
```

to

```java
        private final Queue<Runnable> operations = new ConcurrentLinkedQueue<>();
```

and the relevant call sites.

As such, I'm a little concerned with the exception handling logic in `doWork()` b/c your comments in there make it seem necessary, but in actuality they will never get hit.

Reply:
I might suggest a strongly typed inner class that encapsulates the logic:

```java
private final Queue<SubmitStreamOperation> operations = new ConcurrentLinkedQueue<>();
        
class SubmitStreamOperation {
    ...
    public void run() {
        session.nonExport()
            .onErrorHandler(SendMessageObserver.this::onError)
            ...
    }
}
```

as it makes it easier a bit easier to grok the code paths IMO.

Reply:
Agreed, I'll apply both changes.

Reply:
Fields should also be moved up to the top of the class.